lib/util/stream_utils.js

```diff
@@ -698,6 +703,99 @@ shaka.util.StreamUtils = class {
       }
     } // for (const preferredKeySystem of preferredKeySystems)
 
+    /**
+     * For LR devices, polyfill is used, which calls isTypeSupported and
+     * requestMediaKeySystemAccess instead.
+     * Group similar variants, and share results.
+     * Helpful for video with a lot of variants on LR.
+     */
+    const enableCahe = enableMediaCapabilitiesCache &&
+      !srcEquals && !shaka.util.Platform.isChromecast();
+    if (enableCahe) {
+      const decodeInfoCache = {};
+      const buildCacheKey = (variant) => {
+        if (variant.audio.fullMimeTypes.size < 1 ||
+          variant.video.fullMimeTypes.size < 1) {
+          return undefined;
+        }
+        const audio = variant.audio;
+        const video = variant.video;
+
+        // if pure function for calculating the fulltype
+        const videoDrmInfos = video ? video.drmInfos : [];
+        const audioDrmInfos = audio ? audio.drmInfos : [];
+        const allDrmInfos = videoDrmInfos.concat(audioDrmInfos);
+
+        const keySystems= new Set(allDrmInfos.map((info) => info.keySystem));
+        const cacheKeys = [];
+        for (const keySystem of keySystems) {
+          for (const videoFullMimeType of video.fullMimeTypes) {
+            for (const audioFullMimeType of audio.fullMimeTypes) {
+              cacheKeys.push(
+                  `${videoFullMimeType}#${audioFullMimeType}#${keySystem}`);
+              // video.hdr + audio.spatialAudio + audio.channelsCount
+            }
+          }
+        }
+        return cacheKeys;
+      };
+
+      // sort variants, so that high bandwidth checked first.
+      // High bandwidth tends to be less likely to be supported, or smooth
+      const sortedVariants = variants.sort((a, b) => b.bandwidth - a.bandwidth);
+      const keyToCheck = 'supported';
+
+      for (const variant of sortedVariants) {
+        const cacheKeys = buildCacheKey(variant);
+        const useCache = cacheKeys && cacheKeys.length &&
+            cacheKeys.reduce((hasCache, key) =>
+              (hasCache && decodeInfoCache[key] &&
+                // If not meet criteria, won't use cache.
+                decodeInfoCache[key][keyToCheck])
+            , true);
+        shaka.log.info('getDecodingInfosForVariants used cache.',
+            useCache, cacheKeys, variant);
+        if (useCache) {
+          variant.decodingInfos =
+            cacheKeys.map((key) => decodeInfoCache[key]);
+          continue;
+        }
+
+        /** @type {!Array.<!Array.<!MediaDecodingConfiguration>>} */
+        const decodingConfigs = shaka.util.StreamUtils.getDecodingConfigs_(
```

Comment:
It seems like there is significant code duplication between this code here, and the existing code down below. That sort of code duplication is a maintenance issue, as it would be easy for someone to change one copy of this code but not the other.

So, could I suggest a refactoring?

Break the existing code (from lines 800 to 819) into a new class method. Called `getAndApplyDecodingInfosForVariant_`, or something like that.

Then here, you could do:

```js
await shaka.util.StreamUtils.getAndApplyDecodingInfosForVariant_(
    variant, usePersistentLicenses, srcEquals, preferredKeySystems);
if (cacheKey) {
  for (let i = 0; i < variant.decodingInfos.length; i++) {
    decodeInfoCache[cacheKey] = variant.decodingInfos[i];
    delete decodeInfoCache[cacheKey]['configuration'];
  }
```

Reply:
Make sense, will update accordingly.

Reply:
Fixed.